{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import unittest\n",
    "import os\n",
    "from random import random\n",
    "import checkpoint as ch\n",
    "#del ch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_rand = int(round(random() * 100, 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "class HenryChallenge(unittest.TestCase):\n",
    "    \n",
    "    def test_ListaEnteros_01(self):\n",
    "      lista_test = ch.ListaEnteros(1, 10)\n",
    "      lista_esperada = [1,2,3,4,5,6,7,8,9,10]\n",
    "      self.assertEqual(lista_test, lista_esperada)\n",
    "\n",
    "    def test_ListaEnteros_02(self):\n",
    "      lista_test = ch.ListaEnteros(3, 7)\n",
    "      lista_esperada = [3,4,5,6,7,8,9,10]\n",
    "      self.assertEqual(lista_test, lista_esperada)\n",
    "\n",
    "    def test_ListaEnteros_03(self):\n",
    "      lista_test = ch.ListaEnteros(-2, 4)\n",
    "      lista_esperada = [-2,-1,0,1]\n",
    "      self.assertEqual(lista_test, lista_esperada)\n",
    "\n",
    "    def test_DividirDosNumeros_01(self):\n",
    "      parte_entera, resto = ch.DividirDosNumeros(10, 5)\n",
    "      lista_test = [parte_entera, resto]\n",
    "      lista_esperada = [2,0]\n",
    "      self.assertEqual(lista_test, lista_esperada)\n",
    "\n",
    "    def test_DividirDosNumeros_02(self):\n",
    "      parte_entera, resto = ch.DividirDosNumeros(17, 3)\n",
    "      lista_test = [parte_entera, resto]\n",
    "      lista_esperada = [5,2]\n",
    "      self.assertEqual(lista_test, lista_esperada)\n",
    "\n",
    "    def test_DividirDosNumeros_03(self):\n",
    "      parte_entera, resto = ch.DividirDosNumeros(13, 3)\n",
    "      lista_test = [parte_entera, resto]\n",
    "      lista_esperada = [4,1]\n",
    "      self.assertEqual(lista_test, lista_esperada)\n",
    "\n",
    "    def test_NumeroCapicua_01(self):\n",
    "      valor_test = ch.NumeroCapicua(4589)\n",
    "      valor_esperado = False\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_NumeroCapicua_02(self):\n",
    "      valor_test = ch.NumeroCapicua(92529)\n",
    "      valor_esperado = True\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_NumeroCapicua_03(self):\n",
    "      valor_test = ch.NumeroCapicua('hola')\n",
    "      valor_esperado = None\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_NumeroCapicua_04(self):\n",
    "      valor_test = ch.NumeroCapicua(3333)\n",
    "      valor_esperado = True\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_Factorial_01(self):\n",
    "      valor_test = ch.Factorial(5)\n",
    "      valor_esperado = 120\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_Factorial_02(self):\n",
    "      valor_test = ch.Factorial(1)\n",
    "      valor_esperado = 1\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_Factorial_03(self):\n",
    "      valor_test = ch.Factorial(0)\n",
    "      valor_esperado = None\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_Factorial_03(self):\n",
    "      valor_test = ch.Factorial(10)\n",
    "      valor_esperado = 3628800\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_ProximoPrimo_01(self):\n",
    "      valor_test = ch.ProximoPrimo(5)\n",
    "      valor_esperado = 7\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "      \n",
    "    def test_ProximoPrimo_02(self):\n",
    "      valor_test = ch.ProximoPrimo(61)\n",
    "      valor_esperado = 67\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_ProximoPrimo_03(self):\n",
    "      valor_test = ch.ProximoPrimo(139)\n",
    "      valor_esperado = 149\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_ProximoPrimo_04(self):\n",
    "      valor_test = ch.ProximoPrimo(200)\n",
    "      valor_esperado = None\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_factorizar_numero_01(self):\n",
    "      valor_test = ch.factorizar_numero(5)\n",
    "      valor_esperado = [[5],[1]]\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_factorizar_numero_02(self):\n",
    "      valor_test = ch.factorizar_numero(1428)\n",
    "      valor_esperado = [[2,3,7,17], [2,1,1,1]]\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_factorizar_numero_01(self):\n",
    "      valor_test = ch.factorizar_numero('cinco')\n",
    "      valor_esperado = None\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_ClaseAnimal_01(self):\n",
    "      a = ch.ClaseAnimal('perro','negro')\n",
    "      valor_test = a.CumplirAnios()\n",
    "      valor_test = a.CumplirAnios()\n",
    "      valor_test = a.CumplirAnios()\n",
    "      valor_esperado = 3\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_ClaseAnimal_02(self):\n",
    "      a = ch.ClaseAnimal('ballena','azul')\n",
    "      for i in range(0,10):\n",
    "        valor_test = a.CumplirAnios()\n",
    "      valor_esperado = 10\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_ClaseAnimal_03(self):\n",
    "      a = ch.ClaseAnimal('tortuga','verde')\n",
    "      for i in range(0,100):\n",
    "        valor_test = a.CumplirAnios()\n",
    "      valor_esperado = 100\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_NumeroBinario_01(self):\n",
    "      valor_test = ch.NumeroBinario(5)\n",
    "      valor_esperado = 101\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_NumeroBinario_02(self):\n",
    "      valor_test = ch.NumeroBinario(255)\n",
    "      valor_esperado = 11111111\n",
    "      self.assertEqual(valor_test, valor_esperado)\n",
    "\n",
    "    def test_NumeroBinario_03(self):\n",
    "      valor_test = ch.NumeroBinario(-10)\n",
    "      valor_esperado = None\n",
    "      self.assertEqual(valor_test, valor_esperado)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "test_ClaseAnimal_01 (__main__.HenryChallenge) ... ERROR\n",
      "test_ClaseAnimal_02 (__main__.HenryChallenge) ... ERROR\n",
      "test_ClaseAnimal_03 (__main__.HenryChallenge) ... ERROR\n",
      "test_DividirDosNumeros_01 (__main__.HenryChallenge) ... FAIL\n",
      "test_DividirDosNumeros_02 (__main__.HenryChallenge) ... FAIL\n",
      "test_DividirDosNumeros_03 (__main__.HenryChallenge) ... FAIL\n",
      "test_Factorial_01 (__main__.HenryChallenge) ... FAIL\n",
      "test_Factorial_02 (__main__.HenryChallenge) ... FAIL\n",
      "test_Factorial_03 (__main__.HenryChallenge) ... FAIL\n",
      "test_ListaEnteros_01 (__main__.HenryChallenge) ... FAIL\n",
      "test_ListaEnteros_02 (__main__.HenryChallenge) ... FAIL\n",
      "test_ListaEnteros_03 (__main__.HenryChallenge) ... FAIL\n",
      "test_NumeroBinario_01 (__main__.HenryChallenge) ... FAIL\n",
      "test_NumeroBinario_02 (__main__.HenryChallenge) ... FAIL\n",
      "test_NumeroBinario_03 (__main__.HenryChallenge) ... ok\n",
      "test_NumeroCapicua_01 (__main__.HenryChallenge) ... FAIL\n",
      "test_NumeroCapicua_02 (__main__.HenryChallenge) ... FAIL\n",
      "test_NumeroCapicua_03 (__main__.HenryChallenge) ... ok\n",
      "test_NumeroCapicua_04 (__main__.HenryChallenge) ... FAIL\n",
      "test_ProximoPrimo_01 (__main__.HenryChallenge) ... FAIL\n",
      "test_ProximoPrimo_02 (__main__.HenryChallenge) ... FAIL\n",
      "test_ProximoPrimo_03 (__main__.HenryChallenge) ... FAIL\n",
      "test_ProximoPrimo_04 (__main__.HenryChallenge) ... ok\n",
      "test_factorizar_numero_01 (__main__.HenryChallenge) ... ok\n",
      "test_factorizar_numero_02 (__main__.HenryChallenge) ... "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Resumen\n",
      "Total Tests: 25\n",
      "Total Fallas: 18\n",
      "Total Errores: 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "FAIL\n",
      "\n",
      "======================================================================\n",
      "ERROR: test_ClaseAnimal_01 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 113, in test_ClaseAnimal_01\n",
      "    valor_test = a.CumplirAnios()\n",
      "AttributeError: 'NoneType' object has no attribute 'CumplirAnios'\n",
      "\n",
      "======================================================================\n",
      "ERROR: test_ClaseAnimal_02 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 122, in test_ClaseAnimal_02\n",
      "    valor_test = a.CumplirAnios()\n",
      "AttributeError: 'NoneType' object has no attribute 'CumplirAnios'\n",
      "\n",
      "======================================================================\n",
      "ERROR: test_ClaseAnimal_03 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 129, in test_ClaseAnimal_03\n",
      "    valor_test = a.CumplirAnios()\n",
      "AttributeError: 'NoneType' object has no attribute 'CumplirAnios'\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_DividirDosNumeros_01 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 22, in test_DividirDosNumeros_01\n",
      "    self.assertEqual(lista_test, lista_esperada)\n",
      "AssertionError: Lists differ: [None, None] != [2, 0]\n",
      "\n",
      "First differing element 0:\n",
      "None\n",
      "2\n",
      "\n",
      "- [None, None]\n",
      "+ [2, 0]\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_DividirDosNumeros_02 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 28, in test_DividirDosNumeros_02\n",
      "    self.assertEqual(lista_test, lista_esperada)\n",
      "AssertionError: Lists differ: [None, None] != [5, 2]\n",
      "\n",
      "First differing element 0:\n",
      "None\n",
      "5\n",
      "\n",
      "- [None, None]\n",
      "+ [5, 2]\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_DividirDosNumeros_03 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 34, in test_DividirDosNumeros_03\n",
      "    self.assertEqual(lista_test, lista_esperada)\n",
      "AssertionError: Lists differ: [None, None] != [4, 1]\n",
      "\n",
      "First differing element 0:\n",
      "None\n",
      "4\n",
      "\n",
      "- [None, None]\n",
      "+ [4, 1]\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_Factorial_01 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 59, in test_Factorial_01\n",
      "    self.assertEqual(valor_test, valor_esperado)\n",
      "AssertionError: None != 120\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_Factorial_02 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 64, in test_Factorial_02\n",
      "    self.assertEqual(valor_test, valor_esperado)\n",
      "AssertionError: None != 1\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_Factorial_03 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 74, in test_Factorial_03\n",
      "    self.assertEqual(valor_test, valor_esperado)\n",
      "AssertionError: None != 3628800\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_ListaEnteros_01 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 6, in test_ListaEnteros_01\n",
      "    self.assertEqual(lista_test, lista_esperada)\n",
      "AssertionError: Lists differ: [] != [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]\n",
      "\n",
      "Second list contains 10 additional elements.\n",
      "First extra element 0:\n",
      "1\n",
      "\n",
      "- []\n",
      "+ [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_ListaEnteros_02 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 11, in test_ListaEnteros_02\n",
      "    self.assertEqual(lista_test, lista_esperada)\n",
      "AssertionError: Lists differ: [] != [3, 4, 5, 6, 7, 8, 9, 10]\n",
      "\n",
      "Second list contains 8 additional elements.\n",
      "First extra element 0:\n",
      "3\n",
      "\n",
      "- []\n",
      "+ [3, 4, 5, 6, 7, 8, 9, 10]\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_ListaEnteros_03 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 16, in test_ListaEnteros_03\n",
      "    self.assertEqual(lista_test, lista_esperada)\n",
      "AssertionError: Lists differ: [] != [-2, -1, 0, 1]\n",
      "\n",
      "Second list contains 4 additional elements.\n",
      "First extra element 0:\n",
      "-2\n",
      "\n",
      "- []\n",
      "+ [-2, -1, 0, 1]\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_NumeroBinario_01 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 136, in test_NumeroBinario_01\n",
      "    self.assertEqual(valor_test, valor_esperado)\n",
      "AssertionError: None != 101\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_NumeroBinario_02 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 141, in test_NumeroBinario_02\n",
      "    self.assertEqual(valor_test, valor_esperado)\n",
      "AssertionError: None != 11111111\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_NumeroCapicua_01 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 39, in test_NumeroCapicua_01\n",
      "    self.assertEqual(valor_test, valor_esperado)\n",
      "AssertionError: None != False\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_NumeroCapicua_02 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 44, in test_NumeroCapicua_02\n",
      "    self.assertEqual(valor_test, valor_esperado)\n",
      "AssertionError: None != True\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_NumeroCapicua_04 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 54, in test_NumeroCapicua_04\n",
      "    self.assertEqual(valor_test, valor_esperado)\n",
      "AssertionError: None != True\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_ProximoPrimo_01 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 79, in test_ProximoPrimo_01\n",
      "    self.assertEqual(valor_test, valor_esperado)\n",
      "AssertionError: None != 7\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_ProximoPrimo_02 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 84, in test_ProximoPrimo_02\n",
      "    self.assertEqual(valor_test, valor_esperado)\n",
      "AssertionError: None != 67\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_ProximoPrimo_03 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 89, in test_ProximoPrimo_03\n",
      "    self.assertEqual(valor_test, valor_esperado)\n",
      "AssertionError: None != 149\n",
      "\n",
      "======================================================================\n",
      "FAIL: test_factorizar_numero_02 (__main__.HenryChallenge)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\lopez\\AppData\\Local\\Temp/ipykernel_4788/1456077305.py\", line 104, in test_factorizar_numero_02\n",
      "    self.assertEqual(valor_test, valor_esperado)\n",
      "AssertionError: None != [[2, 3, 7, 17], [2, 1, 1, 1]]\n",
      "\n",
      "----------------------------------------------------------------------\n",
      "Ran 25 tests in 0.022s\n",
      "\n",
      "FAILED (failures=18, errors=3)\n"
     ]
    }
   ],
   "source": [
    "resultado_test = unittest.main(argv=[''], verbosity=2, exit=False)\n",
    "archivo_test = open('resultado_test.csv', 'w')\n",
    "archivo_test.write('Total_Tests,Total_Fallas,Total_Errores\\n')\n",
    "archivo_test.write(str(resultado_test.result.testsRun)+','+str(len(resultado_test.result.failures))+','+str(len(resultado_test.result.errors))+'\\n')\n",
    "archivo_test.close()\n",
    "print('Resumen')\n",
    "print('Total Tests:', str(resultado_test.result.testsRun))\n",
    "print('Total Fallas:', str(len(resultado_test.result.failures)))\n",
    "print('Total Errores:', str(len(resultado_test.result.errors)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "12\n",
    "[[2,3],[2,1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def factorizar_numero(numero):\n",
    "    if (type(numero)!=int):\n",
    "        return None\n",
    "    if (numero < 1):\n",
    "        return None\n",
    "    lista_primos = []\n",
    "    i = 2\n",
    "    while(i < (int(numero/2) + 1)):\n",
    "        primo = True\n",
    "        j = 2\n",
    "        while (j < i):\n",
    "            if (i % j == 0):\n",
    "                primo = False\n",
    "                break\n",
    "            j+=1\n",
    "        if (primo):\n",
    "            lista_primos.append(i)\n",
    "        i+=1\n",
    "    lista_factores = []\n",
    "    lista_exponentes = []\n",
    "    numero_original = numero\n",
    "    for i in lista_primos:\n",
    "        j = 0\n",
    "        while (numero % i == 0):\n",
    "            numero = numero / i\n",
    "            if (j == 0):\n",
    "                lista_factores.append(i)\n",
    "            j+=1\n",
    "        if (j>0):\n",
    "            lista_exponentes.append(j)\n",
    "    if (len(lista_factores) == 0):\n",
    "        return [[numero_original],[1]]\n",
    "    else:\n",
    "        return [lista_factores, lista_exponentes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[2, 3, 7, 17], [2, 1, 1, 1]]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "factorizar_numero(1428)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ClaseAnimal(especie, color):\n",
    "    class Animal:\n",
    "        def __init__(self, especie, color):\n",
    "            self.Especie = especie\n",
    "            self.Color = color\n",
    "            self.Edad = 0\n",
    "\n",
    "        def CumplirAnios(self):\n",
    "            self.Edad += 1\n",
    "            return self.Edad\n",
    "    \n",
    "    a = Animal(especie, color)    \n",
    "    return a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = ClaseAnimal('gato','blanco')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n"
     ]
    }
   ],
   "source": [
    "for i in range(0,10):\n",
    "        valor = a.CumplirAnios()\n",
    "print(valor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def OrdenarDiccionario(mi_diccionario, clave, orden=True):\n",
    "    '''\n",
    "    Esta función recibe como parámetro un diccionario, la clave por la que se quiere ordenar,\n",
    "    y si el ordenamiento es descendente (orden=True) o ascendente (orden=False).\n",
    "    El parámetro orden debe tener el valor True por defecto\n",
    "    Recibe tres argumentos:\n",
    "        mi_diccionario: Será el diccionario a ordenar.\n",
    "        clave: Será la clave del diccionario por la cual se ordene.\n",
    "        orden: True o False, para ordenamiento descendente ó ascendente respectivamente.\n",
    "    Ej:\n",
    "        OrdenarDiccionario({'col1':[1,2,3,4], 'col2':['rojo','azul','blanco','gris']}, 'col2') \n",
    "            Debe retornar {'col1':[4,1,2,3], 'col2':['azul','blanco','gris','rojo']}\n",
    "    '''\n",
    "    return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Comparacion Cadenas de ADN\n",
    "\n",
    "\n",
    "ATGTCTTCCTCGA TGCTTCCTATGAC\n",
    "\n",
    "\n",
    "Para el ejemplo anterior, el resultado es CTTCCT porque que es el conjunto ordenado de bases adyacentes de mayor tamaño que se encuentra en ambas formas de vida."
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "c85384e4cb51c8b72350f3a8712cc8351fdc3955e32a27f9b60c6242ab125f01"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
